mindmeld/augmentation.py

```diff
@@ -241,6 +262,167 @@ def __init__(self, batch_size, language, paths, path_suffix, resource_loader):
+        except zipfile.BadZipfile:
+            logger.error("Unable to extract zip file. Try downloading the model again.")
+        return
+
+    def _prepare_inputs(self, queries):
+        """Processes input as expected by the two different English models
+        Example:
+            The default model requires just <unannotated text>
+            The retain_entities model requires <unannotated text> <EOS> <entity values>
+
+        Args:
+            queries (list(str)): List of application queries.
+
+        Return:
+            model_inputs (list(str)): List of queries to be paraphrased in the
+                                      format required by the models
+            processed_entities (list(ProcessedQuery)): List of ProcessedQuery
+                                                       with entity annotations
+        """
+        model_inputs = []
+        processed_entities = []
+        for query in queries:
+            processed_query = load_query(query,
+                                         query_factory=self._resource_loader.query_factory)
+            if self.retain_entities:
+                text = [processed_query.query.text.strip().lower(), EOS_TOKEN]
+                for entity in processed_query.entities:
+                    text.append(entity.text.lower())
+                processed_entities.append(processed_query)
+                model_inputs.append(' '.join(text))
+            else:
+                model_inputs.append(processed_query.query.text.strip())
+        return model_inputs, processed_entities
+
+    def _replace_with_random_entity(self, paraphrase_text, entity_matches):
+        """Replaces values of annotated entities with randomly sampled ones from gazetteers
+
+            Args:
+                paraphrase_text (str): The paraphrased unannotated text
+                entity_matches (List((Entity,Span))): List of (Entity, Span) values
+                                                      found in the paraphrase_text
+
+            Return:
+                processed paraphrases (ProcessedQuery): ProcessedQuery of the paraphrase_text
+        """
+        new_paraphrase_text = []
+        # Start replacing entities in ascending order of span starts
+        entity_matches.sort(key=lambda x: x[1].start, reverse=False)
+        running_start = 0
+        previous_end = 0
+        replaced_spans_entities = []
+        for (entity, span) in entity_matches:
+            if not running_start:
+                # First entity in the query
+                running_start = span.start
+                # Append text seen so far
+                new_paraphrase_text.append(paraphrase_text[:running_start])
+            else:
+                # Calculate new start based on previously replaced entity length
+                running_start += (span.start - previous_end)
+                # Append text seen between entities
+                new_paraphrase_text.append(paraphrase_text[previous_end:span.start])
```

Comment:
Since value of previous_end in the first iteration is 0, you may consider removing the if block

```suggestion
          # Calculate new start based on previously replaced entity length
          running_start += (span.start - previous_end)
          # Append text before first entity or text seen between entities
          new_paraphrase_text.append(paraphrase_text[previous_end:span.start])
```